avahi-daemon/static-services.c

```diff
@@ -420,6 +449,121 @@ static void XMLCALL xml_start(void *data, const char *el, const char *attr[]) {
     return;
 }
 
+static uint8_t hex(char c)
+{
+  if ((c >= '0') && (c <= '9'))
+    return c - '0';
+  if ((c >= 'A') && (c <= 'F'))
+    return c - 'A' + 10;
+  if ((c >= 'a') && (c <= 'f'))
+    return c - 'a' + 10;
+  return 0xFF;
+}
+
+static int decode_hex_buf(struct xml_userdata *u, uint8_t **out_buf, size_t *out_buf_len) {
+    size_t buf_len = strlen(u->buf);
+    uint8_t *raw_buf;
+    size_t iter;
+    size_t raw_buf_len;
+
+    if (buf_len % 2) {
+        avahi_log_error("%s: parse failure: hex value of the txt record should have an even length", u->group->filename);
+        u->failed = 1;
+        return -1;
+    }
+    raw_buf_len = buf_len / 2;
+    raw_buf = avahi_malloc(raw_buf_len);
+    for (iter = 0; iter < raw_buf_len; iter++) {
+        uint8_t high_nibble = hex(u->buf[iter * 2]);
+        uint8_t low_nibble = hex(u->buf[iter * 2 + 1]);
+
+        if (high_nibble > 0xF || low_nibble > 0xF) {
+            avahi_log_error("%s: parse failure: failed to parse hex data: invalid hex data", u->group->filename);
+            u->failed = 1;
+            avahi_free(raw_buf);
+            return -1;
+        }
+        raw_buf[iter] = (high_nibble << 4) | low_nibble;
+    }
+    *out_buf = raw_buf;
+    *out_buf_len = raw_buf_len;
+    return 0;
+}
+
+static uint8_t base64(char c)
+{
+  if (c >= 'A' && c <= 'Z')
+    return c - 'A';
+  if (c >= 'a' && c <= 'z')
+    return c - 'a' + 26;
+  if (c >= '0' && c <= '9')
+    return c - '0' + 52;
+  if (c == '+')
+    return 62;
+  if (c == '/')
+    return 63;
+  return 255;
+}
+
+static int base64_error(struct xml_userdata *u, uint8_t *raw_buf) {
+    avahi_log_error("%s: parse failure: failed to parse base64 data: invalid base64 data", u->group->filename);
+    u->failed = 1;
+    avahi_free(raw_buf);
+    return -1;
+}
+
+static int decode_base64_buf(struct xml_userdata *u, uint8_t **out_buf, size_t *out_buf_len) {
+    size_t buf_len = strlen(u->buf);
```

Comment:
Actually, `MDEyMwo=` is not base64 encoded value, it is a key `MDEyMwo` that has a base64 encoded empty value. Apparently I forgot to test this case. Thanks for catching this!